src/Databases/DatabaseReplicated.cpp

```diff
@@ -0,0 +1,288 @@
+#include <Databases/DatabaseReplicated.h>
+#include <IO/ReadBufferFromFile.h>
+#include <IO/ReadBufferFromString.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteBufferFromFile.h>
+#include <IO/WriteHelpers.h>
+#include <Interpreters/Context.h>
+#include <Interpreters/executeQuery.h>
+#include <Parsers/queryToString.h>
+#include <Common/Exception.h>
+#include <Common/ZooKeeper/KeeperException.h>
+#include <Common/ZooKeeper/Types.h>
+#include <Common/ZooKeeper/ZooKeeper.h>
+#include <Common/ZooKeeper/Lock.h>
+
+
+namespace DB
+{
+
+
+namespace ErrorCodes
+{
+    extern const int NO_ZOOKEEPER;
+    extern const int LOGICAL_ERROR;
+}
+
+void DatabaseReplicated::setZooKeeper(zkutil::ZooKeeperPtr zookeeper)
+{
+    std::lock_guard lock(current_zookeeper_mutex);
+    current_zookeeper = zookeeper;
+}
+
+zkutil::ZooKeeperPtr DatabaseReplicated::tryGetZooKeeper() const
+{
+    std::lock_guard lock(current_zookeeper_mutex);
+    return current_zookeeper;
+}
+
+zkutil::ZooKeeperPtr DatabaseReplicated::getZooKeeper() const
+{
+    auto res = tryGetZooKeeper();
+    if (!res)
+        throw Exception("Cannot get ZooKeeper", ErrorCodes::NO_ZOOKEEPER);
+    return res;
+}
+
+
+DatabaseReplicated::DatabaseReplicated(
+    const String & name_,
+    const String & metadata_path_,
+    const String & zookeeper_path_,
+    const String & replica_name_,
+    Context & context_)
+    : DatabaseAtomic(name_, metadata_path_, "store/", "DatabaseReplicated (" + name_ + ")", context_)
+    , zookeeper_path(zookeeper_path_)
+    , replica_name(replica_name_)
+{
+    if (!zookeeper_path.empty() && zookeeper_path.back() == '/')
+        zookeeper_path.resize(zookeeper_path.size() - 1);
+    // If zookeeper chroot prefix is used, path should start with '/', because chroot concatenates without it.
+    if (!zookeeper_path.empty() && zookeeper_path.front() != '/')
+        zookeeper_path = "/" + zookeeper_path;
+
+    if (context_.hasZooKeeper()) {
+        current_zookeeper = context_.getZooKeeper();
+    }
+    if (!current_zookeeper)
+    {
+            throw Exception("Can't create replicated database without ZooKeeper", ErrorCodes::NO_ZOOKEEPER);
+    }
+
+    // New database
+    if (!current_zookeeper->exists(zookeeper_path, {}, NULL)) {
+        createDatabaseZKNodes();
+    // Old replica recovery
+    } else if (current_zookeeper->exists(zookeeper_path + "/replicas/" + replica_name, {}, NULL)) {
+        String remote_last_entry = current_zookeeper->get(zookeeper_path + "/replicas/" + replica_name, {}, NULL);
+
+        String local_last_entry;
+        try
+        {
+            ReadBufferFromFile in(getMetadataPath() + ".last_entry", 16);
```

Comment:
it's not necessary, however, having log pointer stored on disks gives the following benefits
1. it shows that disk is alright and we can assume that metadata wasn't corrupted during the inactivity period of a replica.
2. In case that some other replica have chosen the same replica name while a replica was inactive, there's no force actions taken. An admin can choose whether to start used to be inactive replica from scratch using a different name, or kill the replica name new owner and reconfigure names.

Anyway, for now there's no mechanism implemented to assign replica names to nodes and check that assignment afterwards. So a human mistake can easily mess up the cluster.